psbt: Actually use SIGHASH_DEFAULT for PSBT signing

achow101 opened the pull request:
Make the behavior align with the help text by actually using SIGHASH_DEFAULT as the default sighash for signing PSBTs.

achow101 committed:
psbt: Actually use SIGHASH_DEFAULT

Make the behavior align with the help text by actually using
SIGHASH_DEFAULT as the default sighash for signing PSBTs.

MarcoFalke commented:
Is this for backport?

achow101 commented:
> Is this for backport?

There is no explicit PSBT support for Taproot yet, so I don't think so. Although we do still sign PSBTs that have Taproot inputs, but only if we make those inputs final.

GeneFerneau commented:
Concept + code review ACK [96f78e](https://github.com/bitcoin/bitcoin/pull/22514/commits/96f78ed1fdd9c5de2fa5ff125f79b37bd75af9c1)

meshcollider reviewed (commented):
utACK 96f78ed1fdd9c5de2fa5ff125f79b37bd75af9c1

theStack commented on src/wallet/wallet.h:
```diff
@@ -581,7 +581,7 @@ class CWallet final : public WalletStorage, public interfaces::Chain::Notificati
     TransactionError FillPSBT(PartiallySignedTransaction& psbtx,
                   bool& complete,
-                  int sighash_type = 1 /* SIGHASH_ALL */,
+                  int sighash_type = 0 /* SIGHASH_DEFAULT */,
```
Maybe I'm missing something obvious, but any reason why we use magic numbers with comments here (and also for the interface in `src/wallet/scriptpubkeyman.h`), rather than the enum directly?
```suggestion
                  int sighash_type = SIGHASH_DEFAULT,
```

achow101 replied:
There was a time a while ago where that didn't work and this code has just been moved around to various places without changing. I've changed it to use `SIGHASH_DEFAULT` directly.

theStack reviewed (commented):
Concept ACK

achow101 commented:
Apparently without this, we may end up creating final transactions with a lower feerate than expected.

benthecarman reviewed (commented):
crACK ed964019f7af13028c99381c3c2b9b1b4d060409

sipa commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -624,7 +624,7 @@ TransactionError LegacyScriptPubKeyMan::FillPSBT(PartiallySignedTransaction& psb
 
         // Get the Sighash type
-        if (sign && input.sighash_type > 0 && input.sighash_type != sighash_type) {
+        if (sign && input.sighash_type > 0 && sighash_type > 0 && input.sighash_type != sighash_type) {
```
Is this right? I think we should distinguish based on whether the PSBT having having a PSBT_IN_SIGHASH_TYPE record and not. I think the current code treats there not being a record, and the record being 0 as identical?

achow101 replied:
Hmm, yes. I've changed `input.sighash_type` to be an optional. The issue this particular change was test related, so I've updated the tests instead.

achow101 committed:
psbt: Make sighash_type std::optional<int>

It is better to ues an optional to determine whether the sighash type
is set rather than using 0 as a magic number.

Sjors commented:
Mmm, the help text was changed from defaulting to `ALL` to `SIGHASH_DEFAULT` in https://github.com/bitcoin/bitcoin/commit/458a345b0590fd2fa04c7d8d70beb8d57e34bbc8, but that change also made sure for non-taproot it's interpreted as `ALL`.

Let's make the help text say `RPCArg::Default{"DEFAULT for Taproot, ALL otherwise"}` and behave accordingly.

The first commit 62621db looks good to me.

achow101 committed:
rpc: Document that DEFAULT is for Taproot, ALL for everything else

achow101 commented:
> Let's make the help text say `RPCArg::Default{"DEFAULT for Taproot, ALL otherwise"}` and behave accordingly.

Added a commit for that.

Sjors commented:
utACK 322d117f6f39d8534b4ea77cd46fbe0635a7b3c5

Maybe swap the last two commits for clarity (or add "for taproot" to ff1ec7b)

sipa commented on src/psbt.h:
```diff
@@ -86,9 +86,9 @@ struct PSBTInput
 
             // Write the sighash type
-            if (sighash_type > 0) {
+            if (sighash_type != std::nullopt) {
```
In commit "psbt: Make signhash_type std::optional<int>"

Nit: I think the ideomatic way of writing this would be `if (sighash_type.has_value())` or just `if (sighash_type)` (here and further).

sipa commented on test/functional/rpc_psbt.py:
```diff
@@ -455,7 +455,7 @@ def run_test(self):
             for key in signer['privkeys']:
                 wrpc.importprivkey(key)
-            signed_tx = wrpc.walletprocesspsbt(signer['psbt'])['psbt']
+            signed_tx = wrpc.walletprocesspsbt(signer['psbt'], True, "ALL")['psbt']
```
In commit "psbt: Actually use SIGHASH_DEFAULT"

What is this change for?

achow101 replied:
At least one of the tests sets the sighash type in the PSBT. However when we sign, if the sighash type given in the RPC does not match the sighash type in the PSBT, we fail to sign. So this change is to make sure that we do sign those PSBTs.

sipa commented:
utACK 322d117f6f39d8534b4ea77cd46fbe0635a7b3c5

Sjors commented:
re-utACK 3624dc5f974beb0f31800a1828efc46a1dd9f78f after rebase

Sjors commented:
re-utACK c0405ee27fa23a9868f04c052bdc94d7accc57e2

To my surprise, this still worked despite the rename:
```sh
PREV=3624dc5 N=3 && git range-diff `git merge-base --all HEAD $PREV`...$PREV HEAD~$N...HEAD 
```

darosior commented on src/wallet/rpc/spend.cpp:
```diff
@@ -1167,7 +1167,7 @@ RPCHelpMan walletprocesspsbt()
                     {"psbt", RPCArg::Type::STR, RPCArg::Optional::NO, "The transaction base64 string"},
                     {"sign", RPCArg::Type::BOOL, RPCArg::Default{true}, "Also sign the transaction when updating (requires wallet to be unlocked)"},
-                    {"sighashtype", RPCArg::Type::STR, RPCArg::Default{"DEFAULT"}, "The signature hash type to sign with if not specified by the PSBT. Must be one of\n"
+                    {"sighashtype", RPCArg::Type::STR, RPCArg::Default{"DEFAULT for Taproot, ALL otherwise"}, "The signature hash type to sign with if not specified by the PSBT. Must be one of\n"
```
How is it "ALL otherwise"? If a user wants `ALL`, even for non-Taproot descriptor, they now have to explicitly set it. It's a (minor) breaking change which i think should have had a release note.

More confusing is that even if `ALL` is set in the PSBT input, this will use `DEFAULT`. A user has to specify it using the command line for the whole transaction.

achow101 replied:
`ALL` is the sighash type that is signed with by default for non-Taproot things. I don't see how this is incorrect?

darosior replied:
Hmm? That's the point. It's not.

You changed `ParseSighashString` in this PR setting `DEFAULT` for all types, not only Taproot.
https://github.com/bitcoin/bitcoin/blob/19526d937fbd16b3c868f4a7ecd2c076e71eb8fc/src/core_read.cpp#L255-L277

achow101 replied:
`DEFAULT` is an alias for `ALL` internally, and signing non-taproot things will result in a `ALL` being appended to the signature.

darosior replied:
Right, i think the issue was because in my repro i had the sighash flag set in the PSBT input. Have i tried to reproduce it before reporting i could have been more precise, sorry about that.

It's still a (minor) API break since the PSBT was signed without failure in 22.0 but failed in 23.0 (as shown by the necessity to modify the functional test for it to pass in this PR https://github.com/bitcoin/bitcoin/pull/22514/files#r754474555). But since it was released like this already i agree with your comment in #25876 that this doesn't need "fixing" (not sure there could have been any reasonable fix anyways).
